File: demo/src/main/java/com/devbrackets/android/recyclerextdemo/task/OrderUpdateTask.java
package com.devbrackets.android.recyclerextdemo.task;

import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.os.AsyncTask;
import android.util.SparseIntArray;

import com.devbrackets.android.recyclerextdemo.database.ItemDAO;

import java.util.List;

public class OrderUpdateTask extends AsyncTask<Void, Void, Cursor> {
    private SQLiteDatabase database;
    private DBUpdateListener listener;
    private SparseIntArray positionMap;

    /**
     * For simplicity sake I will only perform a single reorder update at a time.
     */
    public OrderUpdateTask(SQLiteDatabase database, SparseIntArray positionMap, DBUpdateListener listener) {
        this.database = database;
        this.positionMap = positionMap;
        this.listener = listener;
    }

    @Override
    protected Cursor doInBackground(Void... params) {
        database.beginTransaction();
        saveUpdates();
        database.setTransactionSuccessful();
        database.endTransaction();

        return ItemDAO.findCursorAll(database);
    }

    @Override
    protected void onPostExecute(Cursor cursor) {
        if (listener != null) {
            listener.onDBUpdated(cursor);
        }
    }

    private void saveUpdates() {
        List<ItemDAO> items = ItemDAO.findAll(database);

        //Iterates through the changed indexes
        for (int i = 0; i < items.size(); i++) {
            int originalPosition = positionMap.get(i, -1);

            //If the item was moved then the value is the original position, and the key the new position
            if (originalPosition != -1) {
                ItemDAO item = items.get(originalPosition);
                if (item != null) {
                    item.setOrder(i);
                    item.save(database);
                }
            }
        }
    }

    public interface DBUpdateListener {
        void onDBUpdated(Cursor cursor);
    }
}

File: demo/src/main/kotlin/com/devbrackets/android/recyclerextdemo/task/OrderUpdateTask.kt
package com.devbrackets.android.recyclerextdemo.task

import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.os.AsyncTask
import android.util.SparseIntArray
import com.devbrackets.android.recyclerextdemo.data.database.ItemDAO

/**
 * An example task for updating the ordering in the database
 */
class OrderUpdateTask(
        private val database: SQLiteDatabase,
        private val positionMap: SparseIntArray,
        private val listener: DBUpdateListener?
) : AsyncTask<Void?, Void?, Cursor?>() {
    override fun doInBackground(vararg params: Void?): Cursor? {
        database.beginTransaction()
        saveUpdates()
        database.setTransactionSuccessful()
        database.endTransaction()
        return ItemDAO.Companion.findCursorAll(database)
    }

    override fun onPostExecute(cursor: Cursor?) {
        listener?.onDBUpdated(cursor)
    }

    /**
     * Iterate through all the items, updating the order column to the new
     * position.
     */
    private fun saveUpdates() {
        val items: List<ItemDAO?> = ItemDAO.Companion.findAll(database)

        //Iterates through the changed indexes
        for (i in items.indices) {
            val originalPosition = positionMap[i, -1]

            //If the item was moved then the value is the original position, and the key the new position
            if (originalPosition != -1) {
                val item = items[originalPosition]
                if (item != null) {
                    item.order = i.toLong()
                    item.save(database)
                }
            }
        }
    }

    interface DBUpdateListener {
        fun onDBUpdated(cursor: Cursor?)
    }
}